Objects/bytesobject.c

```diff
@@ -2810,79 +2810,33 @@ _PyBytes_FromBuffer(PyObject *x)
 }
 
 static PyObject*
-_PyBytes_FromList(PyObject *x)
+_PyBytes_FromSequence(PyObject *x)
 {
-    Py_ssize_t i, size = PyList_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-    _PyBytesWriter writer;
-
-    _PyBytesWriter_Init(&writer);
-    str = _PyBytesWriter_Alloc(&writer, size);
-    if (str == NULL)
-        return NULL;
-    writer.overallocate = 1;
-    size = writer.allocated;
-
-    for (i = 0; i < PyList_GET_SIZE(x); i++) {
-        item = PyList_GET_ITEM(x, i);
-        Py_INCREF(item);
-        value = PyNumber_AsSsize_t(item, NULL);
-        Py_DECREF(item);
-        if (value == -1 && PyErr_Occurred())
-            goto error;
-
-        if (value < 0 || value >= 256) {
-            PyErr_SetString(PyExc_ValueError,
-                            "bytes must be in range(0, 256)");
-            goto error;
-        }
-
-        if (i >= size) {
-            str = _PyBytesWriter_Resize(&writer, str, size+1);
-            if (str == NULL)
-                return NULL;
-            size = writer.allocated;
-        }
-        *str++ = (char) value;
-    }
-    return _PyBytesWriter_Finish(&writer, str);
-
-  error:
-    _PyBytesWriter_Dealloc(&writer);
-    return NULL;
-}
-
-static PyObject*
-_PyBytes_FromTuple(PyObject *x)
-{
-    PyObject *bytes;
-    Py_ssize_t i, size = PyTuple_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-
-    bytes = PyBytes_FromStringAndSize(NULL, size);
+    Py_ssize_t size = PySequence_Fast_GET_SIZE(x);
+    PyObject *bytes = PyBytes_FromStringAndSize(NULL, size);
     if (bytes == NULL)
         return NULL;
-    str = ((PyBytesObject *)bytes)->ob_sval;
-
-    for (i = 0; i < size; i++) {
-        item = PyTuple_GET_ITEM(x, i);
-        value = PyNumber_AsSsize_t(item, NULL);
-        if (value == -1 && PyErr_Occurred())
+    char *s = PyBytes_AS_STRING(bytes);
+    PyObject **items = PySequence_Fast_ITEMS(x);
```

Comment:
This is not thread safe right? (since the argument `x` can be mutated if it is a list).

The original code does not seem thread safe either, so maybe address this in a different PR.

Reply:
We probably just need to throw a big critical section around it.

Reply:
The original code may be not be safe in this aspect, but the size of the list is recomputed every time so we don't necessarily have a crash. In this new code, I think we can have a crash when accessing `item[i]` directly.

(The original code could technically be crashing if `x` gets freed just after the condition has been checked and just before `PyList_GET_ITEM` is called but this should be very rare (or am I overthinking?))

Reply:
Indeed, because of the check in the loop the original code is safe in the normal build, but not in the free-threading build (rare, but possible). As noted earlier one approach is to add a critical section. There is the Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST which covers the cases (list and tuple).

Reply:
Thanks folks for the suggestions. Did not notice that we now have a `Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST` that locks the section only if the sequence is a list. Very cool. I've now enclosed the loop that copies list items in a critical section and made both success and failure cases share the same ending macro through the use of `goto`s. Not sure if there's a better way than using these `goto`s though.